File: parser.py
"""
program -> top_lvl_stmt program
        |  top_lvl_stmt

top_lvl_stmt -> declare
             |  func_dec

declare -> 'var' <identifier> '=' expr ';'

func_dec -> 'function' <identifier> '(' arg_list ')' block
         |  'function' <identifier> '(' ')' block

arg_list -> <identifier> ',' arg_list
         |  <identifier>

block -> '{' stmt_sequence '}'

stmt_sequence -> stmt stmt_sequence
              |  stmt

stmt -> declare
     |   <identifier> '=' expr ';'
     |  'return' expr ';'
     |   modifier ';'
     |   func_call ';'
     |   'if' expr 'then' block 
     |   'if' expr 'then' block 'else' block
     |   'while' expr 'do' block

modifier -> <identifier> '+=' expr
         |  <identifier> '-=' expr
         |  <identifier> '/=' expr
         |  <identifier> '*=' expr
         |  <identifier> '%=' expr

func_call -> <identifier> '(' ')'
          |  <identifier> '(' expr_list ')'

expr_list -> expr ',' expr_list
          |  expr

expr -> and_expr '|' expr
     |  and_expr

and_expr -> comparison '&' and_expr
         |  comparison

comparison -> term == comparison 
           |  term != comparison  
           |  term >  comparison 
           |  term <  comparison 
           |  term >= comparison  
           |  term <= comparison  
           |  term  

term -> factor '+' term
     |  factor '-' term 
     |  factor

factor -> unary '*' factor    
       |  unary '/' factor
       |  unary '%' factor     
       |  unary

unary -> '-' primary
      |  '!' primary
      |  primary

primary -> <int-literal>
        |  <real-literal>
        |  <string-literal>
        |  <bool-literal>
        |  <identifier>
        |  '(' expr ')'
        |  func_call

"""

import sys




# AST node types
class Node:
    def __init__(self):
        self.type = "none"
    

class Program(Node):
    def __init__(self):
        self.type = "Program"
        self.nodes = []

    # consider removing
    def add_top_level_stmt(self, node):
        self.nodes.append( node )

    def value(self):
        return self.type

    def children(self):
        return self.nodes

    def __repr__(self):
        return repr(self.nodes)

class Block(Node):
    def __init__(self, stmts):
        self.type = "Block"
        self.stmts = stmts

    # consider removing
    def add_stmt(self, node):
        self.stmts.append( node )

    def value(self):
        return self.type

    def children(self):
        return self.stmts

    def __repr__(self):
        return repr(self.stmts)

class Assign(Node):
    def __init__(self, identifier, expr_node):
        self.type = "Assign"
        self.identifier = identifier
        self.expr_node = expr_node

    def value(self):
        return "{} {}".format(self.type, self.identifier)

    def children(self):
        return [self.expr_node]

class Declare(Node):
    def __init__(self, identifier, expr_node):
        self.type = "Declare"
        self.identifier = identifier
        self.expr_node = expr_node

    def value(self):
        return "{} {}".format(self.type, self.identifier)

    def children(self):
        return [self.expr_node]

    def __repr__(self):
        return "{} {} = \n{}".format(self.type, self.identifier, repr(self.expr_node))

class FunctionDeclare(Node):
    def __init__(self, identifier, args, body):
        self.type = "FunctionDeclare"
        self.identifier = identifier
        self.args = args
        self.body = body

    def value(self):
        return "{} {} with {}".format(self.type, self.identifier, self.args)

    def children(self):
        return [self.body]

class Return(Node):
    def __init__(self, expr_node):
        self.type = "Return"
        self.expr_node = expr_node

    def value(self):
        return "Return"

    def children(self):
        return [self.expr_node]

class AssignOp(Node):
    def __init__(self, identifier, op, expr_node):
        self.type = "AssignOp"
        self.identifier = identifier
        self.op = op
        self.expr_node = expr_node

    def value(self):
        return "{} {} {}".format(self.type, self.identifier, self.op)

    def children(self):
        return [self.expr_node]

class IfThen(Node):
    def __init__(self, condition, body):
        self.type = "IfThen"
        self.condition = condition
        self.body = body

    def value(self):
        return "If then"

    def children(self):
        return [self.condition] + [self.body]

class IfElse(Node):
    def __init__(self, condition, then_body, else_body):
        self.type = "IfElse"
        self.condition = condition
        self.then_body = then_body
        self.else_body = else_body

    def value(self):
        return "If then else"

    def children(self):
        return [self.condition] + [self.then_body] + [self.else_body]

class While(Node):
    def __init__(self, condition, body):
        self.type = "While"
        self.condition = condition
        self.body = body

    def value(self):
        return "While do"

    def children(self):
        return [self.condition] + [self.body]

class BinaryOp(Node):
    def __init__(self, op, left_expr, right_expr):
        self.type = "BinaryOp"
        self.op = op
        self.left_expr = left_expr
        self.right_expr = right_expr

    def value(self):
        return "{} {}".format(self.type, self.op)

    def children(self):
        return [self.left_expr] + [self.right_expr]

    def __repr__(self):
        return "({}\n left: {}\n right: {}\n)".format(self.op, self.left_expr, self.right_expr)

class UnaryOp(Node):
    def __init__(self, op, expr_node):
        self.type = "UnaryOp"
        self.op = op
        self.expr_node = expr_node

    def value(self):
        return "{} {}".format(self.type, self.op)

    def children(self):
        return [self.expr_node]

    def __repr__(self):
        return "({}\n {})".format(self.op, self.expr_node)

class FunctionCall(Node):
    def __init__(self, identifier, expr_args):
        self.type = "FunctionCall"
        self.identifier = identifier
        self.expr_args = expr_args

    def value(self):
        return "{} {}".format(self.type, self.identifier)

    def children(self):
        return self.expr_args

    def __str__(self):
        return "({}, {}, args: {})".format(self.type, self.identifier, str(self.expr_args))
    

class VariableLookup(Node):
    def __init__(self, identifier):
        self.type = "VariableLookup"
        self.identifier = identifier

    def value(self):
        return "{} {}".format(self.type, self.identifier)

    def children(self):
        return []

    def __str__(self):
        return "({}, {})".format(self.type, self.identifier)

class Literal(Node):
    def __init__(self, literal_type, value):
        self.literal_type = literal_type
        self.literal_value = value

    def value(self):
        return "{} {}".format(self.literal_type, self.literal_value)

    def children(self):
        return []

    def __str__(self):
        return "({}, {})".format(self.value, self.literal_type)

# may seem pointless, but will make the code more readible
def tokenValue(t):
    return t[0]
def tokenType(t):
    return t[1]

# Actual Parsing code begins here
tokens = []
index = 0

# return the current token
def currentToken():
    global tokens
    global index
    if index >= len(tokens):
        return ("", "EOF")
    return tokens[index]

def peekToken(n=1):
    global tokens
    global index
    i = index + n
    if i >= len(tokens):
        return ("", "EOF")
    return tokens[i]

# return the current token, and advance the index
def advance():
    global index
    ct = currentToken()
    index += 1
    return ct


def matchKeyword(kwrd):
    ct = currentToken()
    if ct == (kwrd, "keyword"):
        return advance()
    else:
        parse_error("Expected keyword \'{}\', encountered \'{}\' with value \'{}\'.".format(kwrd, tokenType(ct), tokenValue(ct)))

def matchTokenType(ttype):
    ct = currentToken()
    if tokenType(ct) == ttype:
        return advance()
    else:
        parse_error("Expected token of type \'{}\', encountered \'{}\' with value \'{}\'.".format(ttype, tokenType(ct), tokenValue(ct)))

def matchSymbol(smbl):
    ct = currentToken()
    if ct == (smbl, "symbol"):
        return advance()
    else:
        parse_error("Expected symbol \'{}\', encountered \'{}\' with value \'{}\'.".format(smbl, tokenType(ct), tokenValue(ct)))

def matchLiteral():
    ct = currentToken()
    if tokenType(ct).endswith("literal"):
        return advance()
    else:
        parse_error("Expected literal value, encountered \'{}\' with value \'{}\'.".format( tokenType(ct), tokenValue(ct)))

# basic, but will do for now. Don't feel like adding line number reporting, error recovery until later
def parse_error(msg):
    print("Parse error: {}".format(msg))
    raise SystemExit

def parse_program():
    global tokens
    global index

    ast_root = Program()

    ct = currentToken()
    while tokenType( ct ) != "EOF":
        tv = tokenValue(ct)
        if tv == "var":
            ast_node = parse_declare()
            ast_root.add_top_level_stmt( ast_node )
        elif tv == "function":
            ast_node = parse_function_declare()
            ast_root.add_top_level_stmt( ast_node )
        else:
            parse_error("Encountered unexpected token \'{}\' while parsing top-level statement".format(tv))


        # might be unneeded
        ct = currentToken()

    return ast_root


def parse_function_declare():
    matchKeyword("function")

    fn_name = tokenValue( matchTokenType("identifier") )

    matchSymbol("(")

    if tokenValue( currentToken() ) == ")":
        arg_names = []
    else:
        arg_names = parse_arg_list()
    
    matchSymbol(")")

    body = parse_block()

    return FunctionDeclare(identifier=fn_name, args=arg_names, body=body)

def parse_arg_list():
    arg = tokenValue( matchTokenType("identifier") )

    if tokenValue( currentToken() ) == ",":
        matchSymbol(",")
        return [arg] + parse_arg_list()

    return [arg]

def parse_block():
    # one thing we could do here is accept single statements followed by a semicolon, without requiring braces around it. 
    # this would be a simple matter of checking if the next token was a brace symbol. 
    # I'm undecided on this so far because I'm not sure if I'll like how it looks, and because I'm not sure if I want 
    # to allow single statements in every case where I have a <block> in the grammar. for example, I think it would look
    # strange for function declarations not to include braces.

    matchSymbol("{")
    stmts = [ parse_statement() ]

    # yes I'm doing it iteratively. Functions tend to have no more than 6 or 7 arguments whereas a block can have dozens of statements
    while tokenValue( currentToken() ) != "}":
        stmts.append( parse_statement() )

    matchSymbol("}")
    return Block(stmts=stmts)

def parse_statement():
    ct = currentToken()

    if tokenType(ct) == "identifier":
        nt = peekToken()
        tv = tokenValue(nt)
        if tv == "=":
            stmt = parse_assign()
            return stmt

        if tv == "(":
            stmt = parse_function_call()
            matchSymbol(";")
            return stmt

        return parse_modifier()

    tv = tokenValue(ct)
    
    if tv == "var":
        return parse_declare()

    if tv == "return":
        matchKeyword("return")
        expr = parse_expr()
        matchSymbol(";")
        return Return(expr_node=expr)

    if tv == "if":
        return parse_if()

    if tv == "while":
        return parse_while()

    parse_error("Invalid start of statement, encountered {}".format(tv))

def parse_if():

    matchKeyword("if")
    condition = parse_expr()

    matchKeyword("then")
    then_body = parse_block()

    if tokenValue(currentToken()) == "else":
        matchKeyword("else")
        else_body = parse_block()
        return IfElse(condition=condition, then_body=then_body, else_body=else_body)

    return IfThen(condition=condition, body=then_body)

def parse_while():
    matchKeyword("while")

    condition = parse_expr()

    matchKeyword("do")

    body = parse_block()

    return While(condition=condition, body=body)

def parse_assign():
    identifier = tokenValue( matchTokenType("identifier") )

    matchSymbol("=")

    expression = parse_expr()

    matchSymbol(";")

    return Assign(identifier=identifier, expr_node=expression)


def parse_modifier():
    identifier = tokenValue( matchTokenType("identifier") )
    tv = tokenValue( currentToken() )
    if tv in ["+=", "-=", "/=", "*=", "%="]:
            matchSymbol(tv)

            expr = parse_expr()

            matchSymbol(";")
            return AssignOp(identifier=identifier, op=tv, expr_node=expr)
    
    parse_error("Expected assignment, modifier, or function to follow identifier. Encountered {} {}".format(identifier, tv))

def parse_declare():
    matchKeyword("var")

    identifier = tokenValue( matchTokenType("identifier") )

    matchSymbol("=")

    expression = parse_expr()

    matchSymbol(";")

    return Declare(identifier=identifier, expr_node=expression)

def parse_expr():
    left_expr = parse_and_expr()

    if tokenValue(currentToken()) == "|":
        matchSymbol("|")
        right_expr = parse_expr()
        return BinaryOp(op="|", left_expr=left_expr, right_expr=right_expr)
    
    return left_expr

def parse_and_expr():
    left_expr = parse_comparison()

    if tokenValue(currentToken()) == "&":
        matchSymbol("&")
        right_expr = parse_and_expr()
        return BinaryOp(op="&", left_expr=left_expr, right_expr=right_expr)

    return left_expr

def parse_comparison():
    left_expr = parse_term()
    
    tv = tokenValue(currentToken())
    if tv in ["==", "!=", "==", "<=", ">=", "<", ">"]:
        matchSymbol(tv)
        right_expr = parse_comparison()
        return BinaryOp(op=tv, left_expr=left_expr, right_expr=right_expr)
    else:
        return left_expr

def parse_term():
    left_expr = parse_factor()

    tv = tokenValue(currentToken())
    if tv in ["+", "-"]:
        matchSymbol(tv)
        right_expr = parse_term()
        return BinaryOp(op=tv, left_expr=left_expr, right_expr=right_expr)
    else:
        return left_expr

def parse_factor():
    left_expr = parse_unary()

    tv = tokenValue(currentToken())
    if tv in ["*", "/", "%"]:
        matchSymbol(tv)
        right_expr = parse_factor()
        return BinaryOp(op=tv, left_expr=left_expr, right_expr=right_expr)
    else:
        return left_expr

def parse_unary():
    tv = tokenValue(currentToken())
    if tv in ["+", "-"]:
        matchSymbol(tv)
        expr_node = parse_primary()
        return UnaryOp(op=tv, expr_node=expr_node)
    return parse_primary()

def parse_primary():
    ct = currentToken()

    if tokenValue(ct) == "(":
        matchSymbol("(")
        parenthesized_expr = parse_expr()
        matchSymbol(")")
        return parenthesized_expr

    if tokenType(ct) == "identifier":
        nt = peekToken()
        if tokenValue(nt) == "(":
            return parse_function_call()
        identifier = tokenValue( matchTokenType("identifier") )
        return VariableLookup(identifier=identifier)

    value, literal_type = matchLiteral()
    return Literal(literal_type=literal_type, value=value)


def parse_function_call():
    fn_name = tokenValue( matchTokenType("identifier") )
    matchSymbol("(")

    if tokenValue( currentToken() ) == ")":
        expr_args = []
    else:
        expr_args = parse_expr_list()

    matchSymbol(")")
    return FunctionCall(identifier=fn_name, expr_args=expr_args)

def parse_expr_list():
    expr = parse_expr()

    # of course this could be done iteratively, but if I'm gonna write a recusive descent parser I may as well commit to it
    if tokenValue( currentToken() ) == ",":
        matchSymbol(",")
        return [expr] + parse_expr_list()

    return [expr]
    
### End of parsing code ###

# adapted from https://vallentin.dev/2016/11/29/pretty-print-tree
def pretty_print_ast(node, _prefix="", _last=True):
    print(_prefix, "`- " if _last else "|- ", node.value(), sep="" )
    _prefix += "   " if _last else "|  "
    child_count = len(node.children())
    for i, child in enumerate(node.children()):
        _last = i == (child_count - 1)
        pretty_print_ast(child, _prefix, _last)
    

if __name__ == "__main__":
    
    args = [arg for arg in sys.argv[1:] if not arg.startswith("-")]

    _file = "tokens.txt"
    if len(args) > 0:
        _file = args[0]


    with open(_file, 'r') as file:
        data = file.read().replace('\n', '')
    
    # wow dangerous
    tokens = eval(data)
    #print(tokens)
    ast = parse_program()

    pretty_print_ast(ast)

File: cpp/src/parser.cpp
#include "parser.h"
#include "token.h"
#include "ast.h"

#include <string>
#include <iostream>
#include <memory> 

using std::vector;
using std::string;
using std::shared_ptr;


/*
*   ParseState Methods
*/

Token ParseState::currentToken () {
    if (index >= tokens.size())
        return EOF_TOKEN;
    return tokens[index];
}

Token ParseState::peekToken (int n) {
    int i { n + 1 };
    if ( i >= tokens.size())
        return EOF_TOKEN;
    return tokens[i];
}

Token ParseState::advance() {
    Token ct = currentToken();
    ++index;
    return ct;
}

Token ParseState::matchKeyword(string kwrd) {
    Token ct = currentToken();
    std::string tv = ct.sval;
    TokenType   tt = ct.type;
    
    if ( tt == TokenType::keyword && kwrd == tv )
        return advance();

    std::cout << "Expected keyword " << kwrd <<  " , encountered " 
        << tokenTypeString(tt) << " with value " << tv << std::endl;
    parsingError();
    return EOF_TOKEN;
}

Token ParseState::matchTokenType(TokenType ttype) {
    Token ct = currentToken();

    if ( ct.type == ttype )
        return advance();
    
    std::cout << "Expected token of type " << tokenTypeString(ttype) <<  " , encountered " 
        << tokenTypeString(ct.type) << " with value " << ct.sval << std::endl;
    parsingError();
    return EOF_TOKEN;
}

Token ParseState::matchSymbol(string smbl) {
    Token ct = currentToken();

    if ( ct.type == TokenType::symbol && ct.sval == smbl ) 
        return advance();
    
    std::cout << "Expected symbol " << smbl <<  " , encountered " 
        << tokenTypeString(ct.type) << " with value " << ct.sval << std::endl;
    parsingError();
    return EOF_TOKEN;
}

Token ParseState::matchLiteral() {
    Token ct = currentToken();
    TokenType tt = ct.type;
    if ( tt == TokenType::bool_literal || tt == TokenType::int_literal || tt == TokenType::real_literal || tt == TokenType::string_literal )
        return advance();

    std::cout << "Expected literal value, encountered "  << tokenTypeString(ct.type) 
        << " with value " << ct.sval << std::endl;
    parsingError();
    return EOF_TOKEN;
}

void ParseState::parsingError() {
    std::cout << "Parsing error encountered, terminating." << std::endl;
    exit(-1);
}


/*
*  Recursive-descent parsing methods
*/

shared_ptr<BaseNode> parse_program(ParseState &parse_state) {
    auto ast_root = std::make_shared<Program>();

    Token ct = parse_state.currentToken(); 
    while (ct.type != TokenType::eof) {
        if ( ct.sval == "var" ) {
            auto ast_node = parse_declare(parse_state);
            ast_root->add_top_level_stmt(ast_node);
        }
        else if (ct.sval == "function") {
            auto ast_node = parse_function_declare(parse_state);
            ast_root->add_top_level_stmt(ast_node);
        }
        else {
            std::cout << "Encountered unexpected token " << ct.sval 
                << " while parsing top-level statement." << std::endl;
            parse_state.parsingError();
        }
        ct = parse_state.currentToken();
    }
    return ast_root;
}

shared_ptr<BaseNode> parse_function_declare(ParseState &parse_state) {
    parse_state.matchKeyword( string{"function"} );
    Token identifier_token = parse_state.matchTokenType( TokenType::identifier );
    parse_state.matchSymbol( string{"("} );

    vector<string> arg_names;
    if ( parse_state.currentToken().sval != ")" ) {
        do {
            Token arg = parse_state.matchTokenType( TokenType::identifier );
            arg_names.push_back(arg.sval);
        } 
        while ( 
            // use short-circuiting here to advance the parser state only if the comma is encountered
            ( parse_state.currentToken().sval == "," ) && 
            ( parse_state.advance().type != TokenType::eof ) 
        );
    }

    parse_state.matchSymbol( string{")"} );
    auto body = parse_block(parse_state);
    return std::make_shared<FunctionDeclare>(identifier_token.sval, arg_names, body);
}


shared_ptr<BaseNode> parse_block(ParseState &parse_state) {
    parse_state.matchSymbol( string{"{"} );
    vector<shared_ptr<BaseNode>> stmts { parse_statement(parse_state) };

    while ( parse_state.currentToken().sval != "}" )
        stmts.push_back( parse_statement(parse_state) );

    parse_state.matchSymbol( string{"}"} );
    return std::make_shared<Block>(stmts);
}

shared_ptr<BaseNode> parse_statement(ParseState &parse_state) {
    auto current_token = parse_state.currentToken();

    if ( current_token.type == TokenType::identifier || current_token.sval == "$" ) {
        auto lvalue = parse_primary(parse_state);

        if ( lvalue->type() == NodeType::FunctionCall ) {
            parse_state.matchSymbol(string{";"});
            return lvalue;
        }
        return parse_assignment(parse_state, lvalue);
    }

    shared_ptr<BaseNode> statement;
    if ( current_token.sval == "var" ) { 
        statement =  parse_declare(parse_state);
    }
    else if ( current_token.sval == "if" ) {
        statement = parse_if(parse_state);
    }          
    else if ( current_token.sval == "while" ) {
        statement =  parse_while(parse_state);
    }           
    else if ( current_token.sval == "return" ) {
        parse_state.matchKeyword( string{"return"} );
        auto expr = parse_expr(parse_state);
        parse_state.matchSymbol( string{";"});
        statement = std::make_shared<Return>(expr);
    } 
    else {
        std::cout << "Invalid start of statement, encountered " << current_token.sval << std::endl;
        parse_state.parsingError();
    }

    return statement;
}

shared_ptr<BaseNode> parse_if(ParseState &parse_state){
    parse_state.matchKeyword( string{"if"} );
    auto condition = parse_expr(parse_state);
    parse_state.matchKeyword("then");
    auto then_body = parse_block(parse_state);

    if( parse_state.currentToken().sval == "else"  ) {
        parse_state.matchKeyword("else");
        auto else_body = parse_block(parse_state);
        return std::make_shared<IfElse>(condition, then_body, else_body);
    }
    
    return std::make_shared<IfThen>(condition, then_body);
}

shared_ptr<BaseNode> parse_while(ParseState &parse_state) {
    parse_state.matchKeyword( string{"while"} );
    auto condition = parse_expr(parse_state);
    parse_state.matchKeyword( string{"do"} );
    auto body = parse_block(parse_state);
    return std::make_shared<While>(condition, body);
}

shared_ptr<BaseNode> parse_assignment(ParseState &parse_state, shared_ptr<BaseNode> lvalue) {

    if (lvalue->type() != NodeType::VariableLookup && lvalue->type() != NodeType::Access) {
        std::cout << "Invalid L value:  " << lvalue->value() << std::endl;
        parse_state.parsingError();
    }

    auto token_value = parse_state.currentToken().sval;
    if ( !(token_value == "="  || token_value == "+=" || token_value == "-=" || 
        token_value == "/=" || token_value == "*=" || token_value == "%=" ) ) 
    {
        std::cout << "Expected assignment operator, encountered  " << token_value << std::endl;
        parse_state.parsingError();
    }
    
    parse_state.matchSymbol(token_value);
    auto expr = parse_expr(parse_state);
    parse_state.matchSymbol(";");
    return std::make_shared<AssignOp>(lvalue, token_value, expr);
}

shared_ptr<BaseNode> parse_declare(ParseState &parse_state) {

    parse_state.matchKeyword("var");
    auto id = parse_state.matchTokenType(TokenType::identifier);
    parse_state.matchSymbol("=");
    auto expr = parse_expr(parse_state);
    parse_state.matchSymbol(";");
    return std::make_shared<Declare>(id.sval, expr);
}

shared_ptr<BaseNode> parse_expr(ParseState &parse_state, int rbp) {
    auto left_expr = parse_primary(parse_state);

    // use Pratt parsing for operator precedence in expressions
    while ( binding_power( parse_state.currentToken() ) > rbp ) {
        auto op = parse_state.currentToken();
        parse_state.advance();
        // last arg, right_expr, stores the result of the recursive call here
        left_expr = std::make_shared<BinaryOp>(op.sval, left_expr, parse_expr(parse_state, binding_power(op)));
    }
    return left_expr;
}

shared_ptr<BaseNode> parse_unary(ParseState &parse_state) {
    auto token_value = parse_state.currentToken().sval;
    parse_state.matchSymbol(token_value);
    auto expr_node = parse_primary(parse_state);
    return std::make_shared<UnaryOp>(token_value, expr_node);
}

shared_ptr<BaseNode> parse_primary(ParseState &parse_state) {
    auto current_token = parse_state.currentToken();

    // vector literals
    if ( current_token.sval == "[" || current_token.sval == "<[" ) {
        // parsing both heterogeneous and homogenous vectors the same way, since type enforcement won't happen until later
        auto token_value =  current_token.sval;
        parse_state.matchSymbol(token_value);

        // vector_type unnecessary right now, so commented out for now
        //string vector_type;
        string closing;
        if ( token_value == "[" ) {
            //vector_type = "Heterogeneous";
            closing = "]";
        }
        else {
            //vector_type = "Heterogeneous";
            closing = "]>";   
        }
        auto vector_contents = parse_expr_list(parse_state);
        parse_state.matchSymbol(closing);
        return std::make_shared<VectorLiteral>(vector_contents);
    }

    // unary ops
    if ( current_token.sval == "!" || current_token.sval == "-" ) {
        return parse_unary(parse_state);
    }

    shared_ptr<BaseNode> primary_expr;

    // parenthesized expression
    if ( current_token.sval == "(" ) {
        parse_state.matchSymbol("(");
        auto parenthesized_expr = parse_expr(parse_state);
        parse_state.matchSymbol(")");
        primary_expr = parenthesized_expr;
    }
    // identifier
    else if (current_token.type == TokenType::identifier) {
        auto id = parse_state.matchTokenType(TokenType::identifier).sval;
        primary_expr = std::make_shared<VariableLookup>(id, false);
    }
    // sigiled identifier (global lookup)
    else if ( current_token.sval == "$" ) {
        parse_state.matchSymbol("$");
        auto id = parse_state.matchTokenType(TokenType::identifier).sval;
        primary_expr = std::make_shared<VariableLookup>(id, true);
    }

    if ( primary_expr != nullptr ) {
        // both an identifier or a parenthesized expression could be followed by access brackets or fn call
        auto current_token = parse_state.currentToken();
        while ( true ) {
            if ( current_token.sval == "(" ) {
                auto fn_call_args = parse_function_call(parse_state);
                primary_expr =  std::make_shared<FunctionCall>(primary_expr, fn_call_args);
            }   
            else if ( current_token.sval == "[" ) {
                parse_state.matchSymbol("[");
                auto index_expr = parse_expr(parse_state);
                parse_state.matchSymbol("]");
                primary_expr = std::make_shared<Access>(primary_expr, index_expr);
            }
            else {
                return primary_expr;
            }
            current_token = parse_state.currentToken();
        }
    }

    // if all else fails, we assume it's a literal
    return parse_literal(parse_state);
}

vector<shared_ptr<BaseNode>> parse_function_call(ParseState &parse_state) {
    parse_state.matchSymbol("(");

    vector<shared_ptr<BaseNode>> expr_args;
    if ( parse_state.currentToken().sval != ")" ) {
        expr_args = parse_expr_list(parse_state);
    }
    parse_state.matchSymbol(")");
    return expr_args;
}

vector<shared_ptr<BaseNode>> parse_expr_list(ParseState &parse_state) {
    vector<shared_ptr<BaseNode>> expr_list;

    do {
        auto expr = parse_expr(parse_state);
        expr_list.push_back(expr);
    } while (parse_state.currentToken().sval == ",");

    return expr_list;
}


shared_ptr<BaseNode> parse_literal(ParseState &parse_state) {
    Token literal_token = parse_state.matchLiteral();
    shared_ptr<BaseNode> result;
    switch(literal_token.type) {
        case TokenType::bool_literal:
            {
                // might be overkill, but don't want to fail silently
                bool bool_value = false;
                if ( literal_token.sval == "true") {
                    bool_value = true;
                }
                else if (literal_token.sval == "false") {
                    bool_value = false;
                }
                else {
                    std::cout << "Invalid contents of a boolean literal token : " << literal_token.sval << std::endl;
                    parse_state.parsingError();
                }

                result = std::make_shared<BoolLiteral>(bool_value);
                break;
            }
        case TokenType::int_literal:
            {
                int int_value = std::stoi(literal_token.sval);
                result = std::make_shared<IntLiteral>(int_value);
                break;
            }
        case TokenType::real_literal:
            {
                double double_value = std::stod(literal_token.sval);
                result = std::make_shared<RealLiteral>(double_value);
                break;
            }
        case TokenType::string_literal:
            {
                result = std::make_shared<StringLiteral>(literal_token.sval);
                break;
            }
        default:
            {
                std::cout << "Invalid token type for literal: " << tokenTypeString(literal_token.type) << std::endl;
                parse_state.parsingError();
            }  
    }
    return result;
}


// entry-point for parsing
shared_ptr<BaseNode> parse_tokens(vector<Token> tokens, bool printout) {
    ParseState parse_state { tokens };
    shared_ptr<BaseNode> ast = parse_program(parse_state);
    
    if (printout)
        pretty_print_ast(ast);

    return ast;
}


// operator-precedence lookup 
int binding_power(Token tok) {
    auto tv = tok.sval;
    if ( tv == "|" || tv == "&" ) 
        return 1;
    if ( tv == "==" || tv ==  "!=" || tv ==  "<=" || tv ==  ">=" || tv ==  "<" || tv ==  ">" ) 
        return 3;
    if ( tv == "+" || tv == "-" ) 
        return 4;
    if ( tv == "*" || tv == "/" || tv == "%" ) 
        return 5;
    return -1;
}

// adapted from https://vallentin.dev/2016/11/29/pretty-print-tree
void pretty_print_ast(shared_ptr<BaseNode> node, string _prefix, bool _last) {
    std::cout << _prefix << ( _last ? "`- " : "|- ") << node->value() << std::endl;
    _prefix = _prefix + ( _last ? "   " : "|  " );
    vector<shared_ptr<BaseNode>> children = node->children();
    int child_count = children.size();
    int i = 0;
    for (auto ch : children) {
        _last = ( i == (child_count - 1));
        pretty_print_ast(ch, _prefix, _last);
        ++i;
    }
}

